### ISSUE TITLE
Evaluating partial batches with pmap (via masking) should be easier

### ISSUE THREAD

**Issue by georgedahl** [google/jax#3088]
Currently, if we want to make use of data parallelism and multiple devices to speed up neural network training and evaluation, we use pmap to parallelize the forward pass through the network. This introduces a shape constraint that each device in the pmap must process the same number of examples. In other words, the number of devices must evenly divide the batch size. For training, we can just arrange for our data pipeline to never produce partial batches since we can sample examples in many different similarly effective ways (e.g. sampling with replacement). But for computing validation error and other eval metrics on held out data, we don't control the number of examples in the test set and must process each example exactly once to get the correct numbers.

To handle the case where num_devices does not evenly divide the number of examples in the final partial batch of the test set, we can pad the batch to a batch size that is a multiple of num_devices, and mask out the contribution of the fake examples to the loss computation, being careful to weight the batch's contribution to the aggregate metrics correctly by the number of real examples in the batch. If an automasking transformation could make this easier, this would be really cool.

**Comment by froystig** [comment]
See also #2643 regarding documentation of `mask`.
